### ISSUE TITLE
Plugin request: VGM plugins

### ISSUE THREAD

**Issue by keenanweaver** [ludouzi/fooyin#89]
Shed a tear and quietly uttered "it's beautiful" when I discovered fooyin this morning. Thank you for making this!

My primary use-case for foobar2000 when I used Windows was video game music playback. It has a lot of plugins and is pretty much the de facto way to listen to VGM on Windows. It'd be wonderful if fooyin could replicate that experience on Linux. [Here are the plugins that foobar has](https://www.foobar2000.org/components/tag/game+music).

- [ ] [2SF Decoder](https://www.foobar2000.org/components/view/foo_input_vio2sf)
- [ ] [AdPlug](https://www.foobar2000.org/components/view/foo_input_adplug)
- [ ] [ASAP Decoder](https://www.foobar2000.org/components/view/foo_asap)
- [ ] [foo_input_vgm](https://ux.getuploader.com/foobar2000/) More info: https://vgmrips.net/forum/viewtopic.php?p=5638#p5638
- [ ] [Game Music Emu Decoder](https://www.foobar2000.org/components/view/foo_input_gme)
- [ ] [GSF Decoder](https://www.foobar2000.org/components/view/foo_input_gsf)
- [ ] [Hively Tracker Decoder](https://www.foobar2000.org/components/view/foo_input_hvl)
- [ ] [Lunar 2 PCM Decoder](https://www.foobar2000.org/components/view/foo_lunar2)
- [ ] [MIDI Player](https://www.foobar2000.org/components/view/foo_midi)
- [ ] [NCSF Decoder](https://www.foobar2000.org/components/view/foo_input_ncsf)
- [ ] [OpenMPT Module Decoder](https://www.foobar2000.org/components/view/foo_openmpt54)
- [ ] [PMD Decoder](https://www.foobar2000.org/components/view/foo_input_pmd)
- [ ] [PSF Decoder](https://www.foobar2000.org/components/view/foo_psf)
- [ ] [QSF Decoder](https://www.foobar2000.org/components/view/foo_input_qsf)
- [ ] [SID Decoder](https://www.foobar2000.org/components/view/foo_sid+%28x64%29)
- [ ] [SNESAPU Input](https://www.foobar2000.org/components/view/foo_snesapu)
- [ ] [SSF/DSF Decoder](https://www.foobar2000.org/components/view/foo_input_ht)
- [ ] [USF Decoder](https://www.foobar2000.org/components/view/foo_input_usf)
- [ ] [VGMstream Decoder](https://www.foobar2000.org/components/view/foo_input_vgmstream)
- [ ] [ZXTune Player](https://www.foobar2000.org/components/view/foo_input_zxtune)

The second part of this request is archive playback. Lots of VGM is packaged up in zip, vgm7z/vgz, type formats, for ease of distribution. [Here's an example](https://vgmrips.net/packs/pack/out-run-arcade). Not sure if fooyin already supports that, but being able to add all these different types of archive as files to the playlist would be perfect. I will open up a second request if it is needed.

I and many others would be grateful if these were implemented. I know it's a ton of work.

**Comment by ludouzi** [comment]
Thanks for the request!

Archive support has already been requested by others and so I will definitely be putting it on the roadmap. It's implementation will be similar to CUE sheets with a single disk image, so I'll see about adding it after that.

For decoder plugins, the groundwork for me (or others) to implement them is almost there. Some of the above plugins will be more trivial than others, and it is a lot of work. It's something I'll get around to much later on in development, but it will be done at some point.

**Comment by keenanweaver** [comment]
That's great! Glad to hear that. I know that kode54, the author of many of these plugins, is on Github, and it might be worth reaching out to them for assistance in case you needed it.

**Comment by wardhale** [comment]
This is the single issue keeping me from switching over. It'll be an amazing day when we get the video game format support. VGMstream might be all you need.
https://github.com/vgmstream/vgmstream

**Comment by ludouzi** [comment]
Now that decoder plugins can be implemented, I'll be looking into this very soon. VGMstream can handle a lot of the most common formats from a quick glance, so I'll add a plugin for that first.

**Comment by kddlb** [comment]
@kode54 check this out 👀

**Comment by wardhale** [comment]
Well, this is all awesome. Thanks @ludouzi for looking into it.

**Comment by kode54** [comment]
Sure, I'll look into this, too.

**Comment by kode54** [comment]
Hmm, sure could use a skeleton SDK for dropping projects into, other than gutting the fooyin project tree.

**Comment by ludouzi** [comment]
Yeah, that's on my to-do list along with documentation. I'll have something out by v0.5.4.

**Comment by ludouzi** [comment]
Just thought I'd make a quick progress report.

- **foo_input_vgm (libvgm)**: Initial implementation is finished - playback of vgm, vgz and others is fine. Still needs more work in terms of configuration and handling of duration.
- **OpenMPT Module**: Currently in progress, but handles playback of most formats fine. Need to improve the input api to deal with subsongs for this (and for others).

I've also got working implementations of a VGMStream plugin and a GME plugin, though I'll wait until the input api is settled and subsongs can be handled before taking them further.

**Comment by ludouzi** [comment]
Thanks for testing! The issue with resuming from a paused state should be fixed with https://github.com/fooyin/fooyin/commit/fff62ae518ae6dac2c476e751956a31c0194922c.

Regarding variables, I posted a list in discussions (https://github.com/fooyin/fooyin/discussions/190) for those relating to tracks. The paths in `Settings->Interface->Artwork` are searched in the order listed, so something like `%path%/*.png` at the end may be useful as a catch-all if there's only a single image.

**Comment by keenanweaver** [comment]
Confirmed resolved. Thanks!

And thanks for the link. I think the catch-all approach works just fine with a nice & tidy music directory with proper filenames. Maybe it could be included by default? Assuming there's a regex mechanism in place for case sensitivity and filetypes.

Also, I might have totally imagined the artwork filenaming thing. It seems like only VGMRips & Project-2612 include "album art" with their packs. (I coulda sworn there was art for SNESmusic and the various joshw repos.)

**Comment by kode54** [comment]
You should also look into adding:

```
mainPlr->SetFileReqCallback(RequestFileCallback, NULL);
```

or similar, possibly with context variable if you need it. At the very least, this callback will be asked to provide `yrw801.rom` for OPL-4 VGM files. I chose to do the dirty deed in my plugins and bundle it, but you should probably just provide a config path to load requested files from and read them from there, asking the user to supply their own sample ROMs.

(I'm of the opinion that this ROM should have been piecemeal included as the used portions in the VGM files themselves, since so many of them already bundle game sample assets anyway.)

Example game from vgm rips: Sonyc.

**Comment by ludouzi** [comment]
Thanks for the hint. That's now been implemented with https://github.com/fooyin/fooyin/commit/0096e41b387b318dd20dabe860dcb03b766d47de.

**Comment by keenanweaver** [comment]
Testing https://github.com/fooyin/fooyin/commit/32e6f6b2a8086ae1921630f6f8f7f153deb79ab6

Looks like certain custom metadata isn't being read by fooyin.

Here's a test song: https://archive.org/download/keygen-music/KEYGENMUSiC%20MusicPack/GGHZ/GGHZ%20-%20Doom%203%20BFG%20Edition%20%2B8%20trn.xm

foobar:

![image](https://github.com/user-attachments/assets/1e110f21-98cc-4c7c-93ca-536c79a4fcc3)
![image](https://github.com/user-attachments/assets/8dc5c1ea-1f2f-4462-91c7-400c3394f5fb)

fooyin

![image](https://github.com/user-attachments/assets/c9930934-c6bb-4038-9cc0-82180ea391b3)

**Comment by ludouzi** [comment]
Thanks. Sample and instrument names are now read and stored with https://github.com/fooyin/fooyin/commit/22057aa06e72307f6661ab919f67a0e4e20ae0b8.

> I'm also having an issue where metadata isn't being read when I add incoming tracks from my NAS

If those tracks play fine that that is definitely unusual. I'll do some testing with remote files to see if I can identify any issues.

**Comment by kode54** [comment]
You need these fields:

```
		p_info.info_set_int(field_patterns, mod->get_num_patterns());
		p_info.info_set_int(field_orders, mod->get_num_orders());
		p_info.info_set_int(field_channels, mod->get_num_channels());
		p_info.info_set_int(field_samples, mod->get_num_samples());
		p_info.info_set_int(field_instruments, mod->get_num_instruments());
```

for the counts, and for the names:

```
		std::vector<std::string> _names = mod->get_sample_names();
		...
		_names = mod->get_instrument_names();
		...
		_names = mod->get_channel_names();
		...
		_names = mod->get_pattern_names();
```

And empty testing the iterators as you pass over the lists to skip empty strings, but still incrementing the index number used for the variable name template:
```
		if(name->empty()) continue;
```

**Comment by ludouzi** [comment]
Thanks for the extra info. I'll wait to add the counts until extra property fields can be added to tracks.

On another note, I've decided to implement archive support earlier than planned as it will lead to breaking changes for all audio inputs which will now need to be able to read from memory.

**Comment by keenanweaver** [comment]
Thanks a ton for the archive support!

Main issue I've found is any pack with an `!` fails to load in a playlist. See examples:
https://vgmrips.net/packs/pack/aaahh-real-monsters-mega-drive-genesis
https://vgmrips.net/packs/pack/zoom-mega-drive-genesis

Other special characters like `~` and `&` aren't affected.

And probably as a result of the archives being opened from memory, the 'Open containing folder' function doesn't work on the libarchive inputs.

Other than those, I haven't spotted any other issues!

**Comment by kode54** [comment]
Two of those appear to be due to a decision to use `!` as the separator between archive path and the path within the archive. The other is possibly due to using link references for things like subsong numbers.

Cog implements the # feature for referring to subsongs, but also escapes # in paths using %23, and thus % must also be escaped. Cog instead implements archive URLs similar to foobar2000, in that there's a length field indicating how long the external archive path is. The separator characters are less important when you know the exact length of the field. You could technically still use exclamation points, just you need a length field. Or escaping all special characters in file paths.

**Comment by ludouzi** [comment]
Thanks for the detailed report, and thanks @kode54 for the suggestions. 

Archive paths now use the scheme: `unpack://type|archivelen|file://archive!file`, which resolves all issues relating to `!` (and other special characters).

Open containing folder should also now be working for archive tracks.

The artwork issue was related to how the tracks were updated in the library once they've been 'played'. It was probably because retrieving artwork from archives can be slower than retrieving from file metadata.

**Comment by keenanweaver** [comment]
I noticed some playlist issues that items that use libarchive. It boils down to playlists with archives just don't really work after closing/reopening fooyin.

These M3U playlists don't work:

```
unpack://zip|93|file:///mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan).zip!01 It's the Theme Song!.vgz
unpack://zip|93|file:///mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan).zip!02 It's the Select Screen!.vgz
unpack://zip|93|file:///mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan).zip!03 It's Your Instruction!.vgz
unpack://zip|93|file:///mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan).zip!04 Cooking of Puyo Puyo.vgz
```

```
unpack://zip|48|file:///home/keenan/Music/slowerpace 音楽 - GUILT 罪悪感.zip!slowerpace 音楽 - GUILT 罪悪感 - 01 prelude.flac
unpack://zip|48|file:///home/keenan/Music/slowerpace 音楽 - GUILT 罪悪感.zip!slowerpace 音楽 - GUILT 罪悪感 - 02 hard journey.flac
unpack://zip|48|file:///home/keenan/Music/slowerpace 音楽 - GUILT 罪悪感.zip!slowerpace 音楽 - GUILT 罪悪感 - 03 purification.flac
unpack://zip|48|file:///home/keenan/Music/slowerpace 音楽 - GUILT 罪悪感.zip!slowerpace 音楽 - GUILT 罪悪感 - 04 my ghosts.flac
unpack://zip|48|file:///home/keenan/Music/slowerpace 音楽 - GUILT 罪悪感.zip!slowerpace 音楽 - GUILT 罪悪感 - 05 a real thing.flac
unpack://zip|48|file:///home/keenan/Music/slowerpace 音楽 - GUILT 罪悪感.zip!slowerpace 音楽 - GUILT 罪悪感 - 06 patience.flac
```

While these do:

```
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/01 It's the Theme Song!.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/02 It's the Select Screen!.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/03 It's Your Instruction!.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/04 Cooking of Puyo Puyo.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/05 A Faceoff at the Magician's Tower! ~ Beginning Music.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/06 A Faceoff at the Magician's Tower! ~ Steady, Steady.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/07 A Faceoff at the Magician's Tower! ~ Darkness.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/08 A Faceoff at the Magician's Tower! ~ Quickly Difficult.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/09 Stage Clear.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/10 It's the Ending Theme!.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/11 It's the Staff Roll!.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/12 Ending C.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/13 Warning of Puyo Puyo.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/14 Sorrow of Puyo Puyo.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/15 Victory of Puyo Puyo.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/16 Demo.vgz
/mnt/crusader/Media/Audio/Music/VGM/VGMRips/WonderSwan/Puyo_Puyo_Tsuu_(Bandai_WonderSwan)/17 All Clear.vgz
```

```
slowerpace 音楽 - GUILT 罪悪感/slowerpace 音楽 - GUILT 罪悪感 - 01 prelude.flac
slowerpace 音楽 - GUILT 罪悪感/slowerpace 音楽 - GUILT 罪悪感 - 02 hard journey.flac
slowerpace 音楽 - GUILT 罪悪感/slowerpace 音楽 - GUILT 罪悪感 - 03 purification.flac
slowerpace 音楽 - GUILT 罪悪感/slowerpace 音楽 - GUILT 罪悪感 - 04 my ghosts.flac
slowerpace 音楽 - GUILT 罪悪感/slowerpace 音楽 - GUILT 罪悪感 - 05 a real thing.flac
slowerpace 音楽 - GUILT 罪悪感/slowerpace 音楽 - GUILT 罪悪感 - 06 patience.flac
```

Another small issue: can't seem to save a playlist until I navigate away from it first. In the playlist organizer, the 'save playlist' function will be grayed out until selecting 'default'/'filter results'/any other playlist that have items in them, then I can save it.

**Comment by keenanweaver** [comment]
Latest commits here seemingly fix the reported issues. Thank you! But I found another one that is strange.

I drag-and-drop the [OutRun pack](https://vgmrips.net/packs/pack/out-run-arcade) from my NAS to fooyin, but only three songs get read (the '3D OutRun' tracks.)

If I copy that same pack file to my home directory and try from there, it imports all tracks normally. 

Extracting the pack out and loading that folder on the NAS, all the tracks get imported.

`/home/keenan/Music`
``` bash
.rw-rw-rw- keenan users  719K Wed Aug 26 01:03:00 2015   Out_Run_(Arcade).zip

```

`/mnt/crusader/Media/Audio/Music/VGM/VGMRips/Arcade/SegaSys`
``` bash
.rw-rw-rw- 99 users 719K Wed Aug 26 01:03:00 2015   Out_Run_(Arcade).zip
```

My NAS OS is Unraid, which may do quirky stuff with NFS permissions, but I'm unsure. Just weird it loads only the three tracks. The pack has multiple M3U files in it, but I don't think fooyin does anything with those when importing.

**Comment by ludouzi** [comment]
It's very strange that it manages to load three files from the archive but not the rest. You mentioned issues earlier with the OpenMPT input, so I'm wondering whether this is related. 

I've added some additional debug messages with https://github.com/fooyin/fooyin/commit/8a8a76c5fa77c6d7967e75bce83eeca3cb5d1667 to try and catch any errors when reading archive entries which may or may not help. In the meantime, I'll do some testing with different permissions.

**Comment by kode54** [comment]
I can't reproduce the Archive reading issue with an identical path, but I only tested it as a path on my rootfs. Maybe there's some race condition with opening the same file multiple times at once over a network share with your NAS?

Quick hint for how foobar2000 handles archives: When it's indexing them, it can optionally pre-cache file readers for each individual file that's extracted. The way it reports indexed files meshes well with this, since it uses a callback from the indexing loop. You could pre-index metadata from archive reading another way, I guess.

It shouldn't require unpacking the whole archive to memory at once, since doing one file at a time can also account for the case of archives without solid compression, or other means of storage such as split chunks per group of files.

So perhaps replicating that callback function mechanism, if only for metadata precaching. Probably want to make it optional, of course.

**Comment by ludouzi** [comment]
A race condition is definitely a possibility, especially as it manages to index at least some files in the archive.

Archives are currently handled similar to subsongs in loose files: 

1. Archive reader is initialised, and paths of all readable entries are stored.
2. We get the number of entries, and call ArchiveReader::entry for each.

In the context of LibArchive, we create a new archive reader object for each file, which requires opening the archive and iterating the entries until we get to the file we need. This is where the issue may lie.

A callback for indexing would improve this, as it would mean opening the archive and iterating it's entries only once. I'll see if there's anything useful in the LibArchive API too.

**Comment by keenanweaver** [comment]
Sadly, same behavior on latest master. I'm going to chalk it up to something I'm doing with my setup if no one else can recreate the issue.

**Comment by kode54** [comment]
Quick question, how do I draft up a minimal CMake build script for building a plugin with GUI code? I'm also wondering what I need to add to the CMake scripts to generate the moc source files, like Fooyin does.

**Comment by ludouzi** [comment]
I've added a very simple example widget plugin over at https://github.com/fooyin/fooyin-plugin-examples. That'll be expanded upon with a few more examples over the coming days.

You can use `set(CMAKE_AUTOMOC ON)` to run moc automatically. If you'll be using Qt Designer ui files, you'll also want `set(CMAKE_AUTOUIC ON)`.

**Comment by kode54** [comment]
Feature request for the builtin plugins, and I would like to use it for any I implement as well: Automatically enable the loop-forever mode if Repeat Track is enabled, and disable it again for any other repeat mode, including Repeat Off. This is what I implemented into my DeaDBeeF plugins and into Cog. It also simplifies chasing down Repeat Forever checkboxes in a dozen different plugins any time I feel like listening to one track for a long while, or if I switch back to listening to a series of tracks.

Plus, if you feel like it, for plugins like VGMStream or the OpenMPT one, you could implement reporting the "actual" play position, relative to the current decode offset, if you want to have something like a seekbar that actually jumps back to a relative offset on repeats. Of course, this won't work for emulated formats, since most of them don't have a consistent way of reporting the "actual" play offset within their buried sequence data, or buried streams, or both.

**Comment by GunGunGun** [comment]
Currently Linux desperately needs a good MIDI player, as someone who listen to MIDI music a lot, I find myself.. impossible to enjoy MIDI music on Linux, qmmp doesn't work, drumstick is slow and outputing wierd audio...

I hope someday when fooyin is matured enough we could bring MIDI support.

**Comment by kode54** [comment]
I've started this repo:

https://github.com/kode54/fooyin-kode54-plugins

No packaging yet, just a basic project and working code so far. MIDI using BASSMIDI for now.

**Comment by kode54** [comment]
I have added a branch here:

https://github.com/kode54/fooyin-kode54-plugins/tree/restructure-midi

I literally have no idea what the heck is up with CMake. I don't know how to build a subproject and include it as a dependency without CMake attempting to also build the source files of the same dependencies in the top level project.

**Comment by kode54** [comment]
MIDI working, AUR package `fooyin-kode54-plugins-git` published. Next up, PSF/PSF2/SSF/DSF/QSF/USF/GSF/NCSF/2SF.

**Comment by kode54** [comment]
PSF/PSF2/SSF/DSF/QSF/USF/GSF/2SF/NCSF done. Depends on an update to the SDK to fix a bug with the math.h file from the SDK being included by libmgba.

**Comment by GunGunGun** [comment]
> MIDI working, AUR package `fooyin-kode54-plugins-git` published. Next up, PSF/PSF2/SSF/DSF/QSF/USF/GSF/NCSF/2SF.
> 
> Also, SNESAPU is unlikely unless someone ports the assembly to 64 bit, and ports it to Linux.



> PSF/PSF2/SSF/DSF/QSF/USF/GSF/2SF/NCSF done. Depends on an update to the SDK to fix a bug with the math.h file from the SDK being included by libmgba.

Beauty! I'm gonna test the hell out of it, I've been look for a good MIDI player for Linux for so long, finally my dream is fufilled, I have a massive MIDI library for offline listening but can't enjoy much without this!

Especially it's using BASSMIDI, so far I've seen only fluidsynth for Linux.

**Comment by keenanweaver** [comment]
@kode54 Are you planning to get your work merged upstream or keeping it in your repo?

**Comment by GunGunGun** [comment]
@kode54 Can you give instruction how to build on non-Arch Linux, it's actually pretty hard to get it to work, tried but no luck so far.

**Comment by kode54** [comment]
VGMStream will come, as soon as I figure out how to implement a dynamically generated file name extension list. Because I'd rather query the list from libvgmstream than hand copy the list over and have to bother synchronizing it every time the core library is updated.

I'd also like to see some sort of API exposing the core "Repeat Track" or "Repeat One" mode as a global read-only setting, because all of these plugins which support endlessly repeating files as an option should be able to synchronize on a single setting for that, rather than all of them having separate Play Forever checkboxes in their preferences.

And as it stands, the GME and PSF plugins don't support that behavior at all, while the MIDI one will repeat forever if you set the loop count to 0. I'd like unity with this feature, though.

Unfortunately, the newer VGMStream API does not allow changing this setting at runtime, only at file open time. However, it should take effect on track change or on Repeat One of the current track, if it's changed during playback by twiddling the Repeat settings.

**Comment by NIICKTCHUNS** [comment]
An EXCELLENT foobar2000 plugin that I would love to see being implemented (or at least being know of) is [Game Emu Player (GEP)](https://www.foobar2000.org/components/view/foo_gep) also made by kode54

This plugin makes it possible to listen to VGMs too, but my main use for this plugin was to listen and being able to export each channel of a VGM individually, for example, Mega Drive had 6 Yamaha synthesizers and 3 SEGA PSGs

<img width="642" height="437" alt="Image" src="https://github.com/user-attachments/assets/09e43e1f-108d-44d7-b63a-799e48a72763" />


With this plugin I can uncheck any audio channel that I want and export the music just fine only with the channels that I checked, this also works for other consoles like SNES and etc.. Would be really cool if any of these features could be implemented on fooyin!

**Comment by kode54** [comment]
GME is already built-in, just with no channel muting control.

**Comment by NIICKTCHUNS** [comment]
> GME is already built-in, just with no channel muting control.

Yeah I know, I can already play VGMs on fooyin, but the my point is the ability to mute individual channels just like GEP could do, this would be very nice to have on fooyin (probably a niche feature but it would be really cool)

**Comment by TSnatch72** [comment]
> > GME is already built-in, just with no channel muting control.
> 
> Yeah I know, I can already play VGMs on fooyin, but the my point is the ability to mute individual channels just like GEP could do, this would be very nice to have on fooyin (probably a niche feature but it would be really cool)

I second this, it's such a nice feature that let's you appreciate the tracks in new ways you wouldn't before

**Comment by kode54** [comment]
For your use case of making stems, you may wish to look at this project I started over 7 years ago, but haven't really touched since then.

Someone else on Github mirrored it, then this person has so far added 36 commits on top of the old code to add features.

https://github.com/maxim-zhao/multidumper

It will use one thread per stem, to spam as much of your CPU power as possible and speed up the process.

**Comment by SeongGino** [comment]
Since Kode's repo doesn't seem to have issues enabled, figured I'd mention that it seems the Highly Experimental and Theoretical backends/MiniPSF&SSF&DSF (and I suspect Quixotic/QSF) playback specifically are a bit broken (immediately exits playback without warning), though the other individual backends (PSF2, 2SF, USF, GSF) and VGMStream still seem to be working just fine.

Least, that's the case with current Arch anyways.

*Other than that though* it's covering virtually everything I've personally wanted, though having it integrated into Fooyin (or at least its Flatpak) would make it more accessible for non-Arch's. :D

**Comment by kode54** [comment]
It doesn't support archives, only raw files on disk.

**Comment by kode54** [comment]
User replied then deleted their reply, asking if I provide release builds. I provide no builds. They're entirely specific to the Qt version on the distribution that builds them.

Edit: Scratch that, user was asking about the VGM player. That's bundled with the player, but I don't know about any release builds on the Github. I've always built it myself.

**Comment by mschwendt** [comment]
> video game music

Just in case, for TFMX, Hippel and Future Composer, there is [this](https://github.com/mschwendt/libtfmxaudiodecoder). It would benefit from a subsong feature, though.

**Comment by arrowgent** [comment]
fooyin --version
fooyin 0.12.6
(previously 0.11.1)

i have a large collection of VGM/VGZ music files that i use

i have tested adding a few of them to fooyin playlist, but no audio plays

does fooyin use `libgme` to play vgm?
```
libgme0/resolute,resolute,resolute,resolute,now 0.6.4-1 amd64 [installed]
  Playback library for video game music files - shared library
```
https://github.com/libgme/game-music-emu

investigation:
```
ldd $(which fooyin)|grep gme
	libgme.so.0 => /usr/lib/x86_64-linux-gnu/libgme.so.0 (0x0000733a02d4d000)
```

`fy.engine: Output initialised: "44100Hz 2ch 16 bit (signed) [FL,FR]" (device: "default" )`



--
in the playlist:
clicking properties shows the proper information for the VGM file
artist, album, title, duration, codec.

fooyin Output device: pipewire
setting to ALSA or SDL (pulseaudio) does not play audio
*apply, close fooyin, relaunch fooyin*

fooyin Playback Decoding: Game Music EMU is checked

MP3, OGG, & FLAC audio plays fine.


--
deadbeef plays the vgm just fine

i typically use libvgm/vgmplay to listen in a terminal.
https://github.com/ValleyBell/libvgm
https://github.com/ValleyBell/vgmplay-libvgm

**Comment by ludouzi** [comment]
The GME plugin included in fooyin handles VGM/VGZ, but GME itself only supports a small number of sound chips.

You'll want to install the libvgm plugin for broader support:

https://github.com/fooyin/fooyin-plugin-libvgm

**Comment by arrowgent** [comment]
got it.  and `fyplugin_vgmstream.so` can be used for other video game formats

great.  this works

this possibly also explains why Audacious' GME plugin doesnt play audio also
`Console music decoder engine based on Game_Music_Emu`


my years, they have listened!
thanks
